lib/widgets/emoji_reaction.dart

```diff
@@ -360,3 +364,234 @@ class _TextEmoji extends StatelessWidget {
       text);
   }
 }
+
+
+void showEmojiPickerSheet({required BuildContext context, required Message message}) {
+  final store = PerAccountStoreWidget.of(context);
+
+  showModalBottomSheet<void>(
+    context: context,
+    // Clip.hardEdge looks bad; Clip.antiAliasWithSaveLayer looks pixel-perfect
+    // on my iPhone 13 Pro but is marked as "much slower":
+    //   https://api.flutter.dev/flutter/dart-ui/Clip.html
+    clipBehavior: Clip.antiAlias,
+    useSafeArea: true,
+    isScrollControlled: true,
+    builder: (BuildContext _) {
+      return SafeArea(
+        minimum: const EdgeInsets.only(bottom: 16),
+        // For _EmojiPickerItem, and RealmContentNetworkImage used in ImageEmojiWidget.
+        child: PerAccountStoreWidget(
+          accountId: store.accountId,
+          child: EmojiPicker(pageContext: context, message: message)));
+    });
+}
+
+class EmojiPicker extends StatefulWidget {
+  const EmojiPicker({
+    super.key,
+    required this.pageContext,
+    required this.message,
+  });
+
+  final BuildContext pageContext;
+  final Message message;
+
+  @override
+  State<EmojiPicker> createState() => _EmojiPickerState();
+}
+
+class _EmojiPickerState extends State<EmojiPicker> with PerAccountStoreAwareStateMixin<EmojiPicker> {
+  late List<EmojiCandidate> availableEmojiCandidates;
+  List<EmojiCandidate>? searchResults;
+
+  @override
+  void onNewStore() {
+    final store = PerAccountStoreWidget.of(context);
+    final allEmojiCandidates = store.allEmojiCandidates();
+
+    bool isPopularUnicodeEmoji(EmojiDisplay candidateEmojiDisplay) {
+      if (candidateEmojiDisplay is! UnicodeEmojiDisplay) return false;
+      return popularUnicodeEmojis.any(
+        (emoji) => emoji.emojiUnicode == candidateEmojiDisplay.emojiUnicode);
+    }
+
+    bool isSelfVoted(EmojiCandidate candidate) {
+      return widget.message.reactions?.aggregated.any((reactionWithVotes) =>
+        reactionWithVotes.reactionType == candidate.emojiType
+          && reactionWithVotes.emojiCode == candidate.emojiCode
+          && reactionWithVotes.userIds.contains(store.selfUserId)
+      ) ?? false;
+    }
+
+    availableEmojiCandidates = allEmojiCandidates
+      .where((candidate) =>
+        !isSelfVoted(candidate) && !isPopularUnicodeEmoji(candidate.emojiDisplay))
+      .toList(growable: false);
+    searchResults = null;
+  }
+
+  // Adapted from web/src/emoji_picker.ts
+  void _filterEmojis(String query) {
```

Comment:
I'd like to have this use the emoji autocomplete machinery instead, so it works the same as in the compose box.

That has some useful features this version doesn't; and I think it's also useful just for them to feel consistent with each other for the user.

Reply:
For excluding the popular and already-reacted emoji, I think a good approach would be to extend EmojiAutocompleteQuery so it takes a parameter with a set of emoji to exclude.

But I'd also be fine with not handling that exclusion in this PR, and making it an M5 or M6 follow-up task. I think it's pretty OK if those emoji appear in the list.